backend-shared/hack/service_account.go

```
@@ -0,0 +1,170 @@
+					if err != nil {
+						return "", err
+					}
+				}
+			}
+		}
+	}
+	return GetServiceAccountBearerToken(clientset, ns, ArgoCDManagerServiceAccount)
+}
+
+// GetServiceAccountBearerToken will attempt to get the provided service account until it
+// exists, iterate the secrets associated with it looking for one of type
+// kubernetes.io/service-account-token, and return it's token if found.
+func GetServiceAccountBearerToken(clientset kubernetes.Interface, ns string, sa string) (string, error) {
+	var serviceAccount *corev1.ServiceAccount
+	var secret *corev1.Secret
+	var err error
+	err = wait.Poll(500*time.Millisecond, 30*time.Second, func() (bool, error) {
+		serviceAccount, err = clientset.CoreV1().ServiceAccounts(ns).Get(context.Background(), sa, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+		// Scan all secrets looking for one of the correct type:
+		for _, oRef := range serviceAccount.Secrets {
+			var getErr error
+			secret, err = clientset.CoreV1().Secrets(ns).Get(context.Background(), oRef.Name, metav1.GetOptions{})
+			if err != nil {
+				return false, fmt.Errorf("Failed to retrieve secret %q: %v", oRef.Name, getErr)
+			}
+			if secret.Type == corev1.SecretTypeServiceAccountToken {
+				return true, nil
+			}
+		}
+		return false, nil
+	})
+	if err != nil {
+		return "", fmt.Errorf("Failed to wait for service account secret: %v", err)
+	}
+	token, ok := secret.Data["token"]
+	if !ok {
+		return "", fmt.Errorf("Secret %q for service account %q did not have a token", secret.Name, serviceAccount)
+	}
+	return string(token), nil
+}
+
+func generateClientFromClusterServiceAccount(hostname string, bearerToken string) client.Client {
+	config := &rest.Config{
+		Host:        hostname,
+		BearerToken: bearerToken,
+	}
+	clientObj, err := client.New(config, client.Options{Scheme: scheme.Scheme})
+	if err != nil {
+		log.Fatal(err)
```

This should return the error, rather than logging and panic-ing.

---

Thanks for spotting this, I missed it during my review. Ouch!